Zend/zend_execute.c

```diff
@@ -1874,13 +1888,23 @@ static zend_always_inline void zend_fetch_dimension_address_read(zval *result, z
 			}
 		}
 	} else {
-		if (type != BP_VAR_IS && UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
-			zval_undefined_cv(EG(current_execute_data)->opline->op1.var, EG(current_execute_data));
+		if (type != BP_VAR_IS) {
+			if (UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
+				zval_undefined_cv(EG(current_execute_data)->opline->op1.var, EG(current_execute_data));
+			} else if (!zend_fetch_dimension_address_read_valid_container(container, support_strings)) {
+				zend_error(E_WARNING, "Variable of type %s does not accept array offsets", zend_get_type_by_const(Z_TYPE_P(container)));
+			}
 		}
 		if (/*dim_type == IS_CV &&*/ UNEXPECTED(Z_TYPE_P(dim) == IS_UNDEF)) {
 			zval_undefined_cv(EG(current_execute_data)->opline->op2.var, EG(current_execute_data));
 		}
-		ZVAL_NULL(result);
+
+		/* return error if we're continuing looking up.  This is OK since after FETCH_DIM_R there will be a FREE op */
+		if ((EG(current_execute_data)->opline + 1)->opcode == ZEND_FETCH_DIM_R) {
+			ZVAL_ERROR(result);
```

Comment:
I concur (just pushed the idea) that this isn't necessarily the best solution.  I'm new to internals, so not totally well versed with all aspects of the executor.  I was trying to see if I couldn't somehow compare the result zval with the next-op-cdodes op1, but the memory never lined up, so I assume there was a zval copy happening somewhere during the vm execution.  Checking if the next op is a fetch + the op1 type is a var, I think, I can assume it's chained on me.  I'm not sure if there's some weird syntaxmagic that could generate a FETCH_DIM_R with an op1 of VAR which isn't part of a chain.